### ISSUE TITLE
[Bug] Inferred types break with certain target types

### ISSUE THREAD

**Issue by C-Higgins** [nobrainr/morphism#148]
When the target type is certain combinations of union types, I think when the unions share a property, the type of the ActionFunction breaks and no longer knows the types of the parameters.

** Reproduce **

```typescript
const test = createSchema<{a:string} | {a:string; b:string},{c:string}>({
  a: it => {} // Parameter 'it' implicitly has 'any' type
  b: it => {} // here 'it' is the correct type
})
```

As you can see, `it` should be of type `{c:string}`, for the first property, but it is not. The other parameters are also any. If you remove `a` from the second type, it works as expected.

**Comment by emyann** [comment]
@C-Higgins Thank for taking the time to report this! Sounds like a weird narrowing/widening issue with Typescript Union Type being processed recursively by:

```ts
export type StrictSchema<Target = any, Source = any> = {
  /** `destinationProperty` is the name of the property of the target object you want to produce */
  [destinationProperty in keyof Target]:
    | ActionString<Source>
    | ActionFunction<Target, Source, Target[destinationProperty]>
    | ActionAggregator<Source>
    | ActionSelector<Source, Target>
    | StrictSchema<Target[destinationProperty], Source>;

export interface ActionFunction<D = any, S = any, R = any> {
  (iteratee: S, source: S[], target: D): R;
}
```

Let me dig this a bit, and come back to you when I have a clue :)

**Comment by emyann** [comment]
@C-Higgins This issue should have been fixed in #149 which is version `1.12.2`, please let me know if you encounter any issues of this type :)
